pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,502 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package privilegedaccessmanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/privilegedaccessmanager/v1alpha1"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	gcp "cloud.google.com/go/privilegedaccessmanager/apiv1"
+	privilegedaccessmanagerpb "cloud.google.com/go/privilegedaccessmanager/apiv1/privilegedaccessmanagerpb"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "privilegedaccessmanager-controller"
+	serviceDomain = "//privilegedaccessmanager.googleapis.com"
+)
+
+type ContainerType int
+
+const (
+	Project ContainerType = iota
+	Folder
+	Organization
+)
+
+func (c ContainerType) String() string {
+	return [...]string{"Project", "Folder", "Organization"}[c]
+}
+
+func init() {
+	registry.RegisterModel(krm.PrivilegedAccessManagerEntitlementGVK, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &entitlementModel{config: *config}, nil
+}
+
+var _ directbase.Model = &entitlementModel{}
+
+type entitlementModel struct {
+	config config.ControllerConfig
+}
+
+func (m *entitlementModel) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building PrivilegedAccessManager client for Entitlement: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *entitlementModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+
+	obj := &krm.PrivilegedAccessManagerEntitlement{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
```

Comment:
Some general suggestion from ln92-215 and external ref:
We can treat projectID, orgID and folderID the same thing in externalRef

pseudo code
```go
func oneOfParent(obj) (string, error) {
   p1 := obj.Spec.ProjectRef
   p2 := obj.Spec.FolderRef
   p3 := obj.Spec.OrgRef 
   if p1 == nil &&  p2 == nil && p3 ==nil {
      return "",  fmt.Errorf()
   }
   if p2 == nil && p3 ==nil {
        project := ResolveProject(p1)
        return "projects/" + project.ID,   nil
   }
   if p1 == nil && p3 ==nil {
        folder := ResolveFolder(p2)
        return "folders/" + folder.ID, nil
   }
   if p1 ==nil && p2 == nil {
        org := ResolveOrg(p3)
         return "organizations/" + org.ID, nil
   }
   return "", fmt.Errorf()
}

func AdapterForObject() {
    parent, err := oneOfParent(obj)
    if ... {
         buildID(parent, resourceID) 
    } else {
          asID(parent)
    }
}

func buildID(parent, resourceID string) {
      return ID{
           Parent: parent,
           ResourceID: resourceID
       }
}

func ValidateExternal(obj *krm.PAM, externalRef *ExternalRef) error {
      parent, err := oneOfParent(obj)
      if parent != externalRef.Parent {
          return err
      }
}
 

```

Reply:
Good point. Done.